src/Resolvers/Microsoft.NET.Sdk.WorkloadMSBuildSdkResolver/WorkloadPartialResolver.cs

```diff
@@ -0,0 +1,258 @@
+                    }
+                }
+            }
+            return new NullResolutionResult();
+        }
+
+        public ResolutionResult Resolve(string sdkReferenceName, string dotnetRootPath, string sdkVersion
+#if USE_SERILOG
+            , List<Action<Serilog.ILogger>> logActions
+#endif
+            )
+        {
+            if (!_enabled)
+            {
+                return new NullResolutionResult();
+            }
+
+            ResolutionResult resolutionResult;
+
+            lock (_lockObject)
+            {
+                if (_cachedState == null ||
+                    _cachedState.DotnetRootPath != dotnetRootPath ||
+                    _cachedState.SdkVersion != sdkVersion)
+                {
+#if USE_SERILOG
+                    if (_cachedState == null)
+                    {
+                        logActions.Add(log => log.Information($"Creating initial {nameof(WorkloadPartialResolver)} state"));
+                    }
+                    else
+                    {
+                        logActions.Add(log => log.Information($"Recreating {nameof(WorkloadPartialResolver)} state (resolved SDK changed)"));
+                    }
+#endif
+
+                    var workloadManifestProvider = new SdkDirectoryWorkloadManifestProvider(dotnetRootPath, sdkVersion);
+                    var workloadResolver = WorkloadResolver.Create(workloadManifestProvider, dotnetRootPath, sdkVersion);
+
+                    _cachedState = new CachedState()
+                    {
+                        DotnetRootPath = dotnetRootPath,
+                        SdkVersion = sdkVersion,
+                        ManifestProvider = workloadManifestProvider,
+                        WorkloadResolver = workloadResolver
+                    };
+                }
+                else
+                {
+#if USE_SERILOG
+                    logActions.Add(log => log.Information($"Using cached {nameof(WorkloadPartialResolver)} state"));
+#endif
+                }
+
+                if (!_cachedState.CachedResults.TryGetValue(sdkReferenceName, out resolutionResult))
+                {
+#if USE_SERILOG
+                    logActions.Add(log => log.Information($"Resolving workload in {nameof(WorkloadPartialResolver)}"));
+#endif
+                    resolutionResult = Resolve(sdkReferenceName, _cachedState.ManifestProvider, _cachedState.WorkloadResolver);
+
+                    _cachedState = _cachedState with
+                    {
+                        CachedResults = _cachedState.CachedResults.Add(sdkReferenceName, resolutionResult)
+                    };
+                }
+#if USE_SERILOG
+                else
+                {
+                    logActions.Add(log => log.Information($"Using cached workload resolution result in {nameof(WorkloadPartialResolver)}"));
+                }
+#endif
+            }
+
+            return resolutionResult;
+        }
+    }
+}
+
+#if !NET
```

Comment:
What is this?

Reply:
This is an attribute which allows using init-only properties on .NET Framework.